### PR TITLE
Add new momentum module

### LINKED ISSUE #494: Refactoring: write new momentum encoder (make it more flexible)
The momentum encoder should be

- not a Mixin anymore
- be flexible in terms of backbone, projection/prediction heads
- This includes supporting multiple projection/prediction heat combine, e.g. for SimSiam

A very good baseline for this is provided in https://github.com/lightly-ai/lightly/blob/483-SSL-model-refactoring-remove-models%2C-provide-full-examples-instead-me/lightly/models/modules/momentum_encoder.py

Furthermore, add a deprecation warning to the olde momentum encoder.

### PR DESCRIPTION
**Pull request by IgorSusmelj**
### Changes

closes #494

- Add new building blocks for momentum encoders (MoCo and BYOL)
- Add sample code for using the new MoCo and BYOL models in the cifar10 benchmark

**Todo:**
- [ ] Update ImageNette benchmark code

### CODE REVIEW

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 1]
```diff
@@ -139,40 +141,64 @@ def get_data_loaders(batch_size: int):
...
+        deactivate_requires_grad(self.projection_head_momentum.parameters())
+
         # create our loss with the optional memory bank
         self.criterion = lightly.loss.NTXentLoss(
             temperature=0.1,
             memory_bank_size=memory_bank_size)
             
     def forward(self, x):
-        self.resnet_moco(x)
+        x = self.backbone(x)
```
Comment (philippmwirth) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:170]: You probably need the flatten here too

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 2]
```diff
@@ -255,9 +281,67 @@ def configure_optimizers(self):
...
+        super().__init__(dataloader_kNN, num_classes)
+        # create a ResNet backbone and remove the classification head
+        resnet = torchvision.models.resnet18()
+        self.backbone = nn.Sequential(
+            *list(resnet.children())[:-1],
+            nn.AdaptiveAvgPool2d(1)
+        )
+
+        # create a byol model based on ResNet
+        self.projection_head = BYOLProjectionHead(512,1024,256)
```
Comment (philippmwirth) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:295]: Please put spaces between the numbers

File: docs/source/getting_started/benchmarks/cifar10_benchmark.py [hunk 3]
```diff
@@ -255,9 +281,67 @@ def configure_optimizers(self):
...
+        self.backbone_momentum = copy.deepcopy(self.backbone)
+        self.projection_head_momentum = copy.deepcopy(self.projection_head)
+
+        deactivate_requires_grad(self.backbone_momentum.parameters())
+        deactivate_requires_grad(self.projection_head_momentum.parameters())
+
+        self.criterion = lightly.loss.SymNegCosineSimilarityLoss()
+
+    def forward(self, x):
+        x = self.backbone(x)
```
Comment (philippmwirth) [inline docs/source/getting_started/benchmarks/cifar10_benchmark.py:307]: Also here don't forget the flatten

File: lightly/models/modules/__init__.py [hunk 4]
```diff
@@ -17,3 +17,8 @@
 from lightly.models.modules.heads import SimSiamProjectionHead
 from lightly.models.modules.heads import SimSiamPredictionHead
 from lightly.models.modules.nn_memory_bank import NNMemoryBankModule
+from lightly.models.modules.momentum import batch_shuffle
```
Comment (philippmwirth) [inline lightly/models/modules/__init__.py:20]: None of these are modules, they are functions. I would not put them here. Maybe we can have something like `lightly.functionals` or put them under `lightly.utils`.

Comment (philippmwirth) [inline lightly/models/modules/__init__.py:20]: None of these are modules, they are functions. I would not put them here. Maybe we can have something like `lightly.functionals` or put them under `lightly.utils`.

File: lightly/models/modules/momentum.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        >>> out0 = projection_head_momentum(f0)
+        >>> out1 = batch_unshuffle(out1, shuffle)
+    """
+    unshuffle = torch.argsort(shuffle)
+    return batch[unshuffle]
+
+def deactivate_requires_grad(params):
+    """Deactivates the requires_grad flag for all parameters.
+    
+    This has the same effect as executing the model within a `torch.no_grad()`
```
Comment (philippmwirth) [inline lightly/models/modules/momentum.py:43]: I would mention that it's like permanently executing the model within a `torch.no_grad()` context. 

Small typo: "Use this method to disable**d** gradient computation..."

**Review by philippmwirth — CHANGES_REQUESTED** [review]
Great job although I have some feedback.

Also, can you please add tests so that coverage doesn't drop?

File: lightly/models/modules/__init__.py (hunk 4 above)
Comment (philippmwirth) [reply lightly/models/modules/__init__.py:20]: After the discussion with @IgorSusmelj we have another idea: Put it in `lightly.models.utils(.py)` or put it in `lightly.models.momentum.py`

**Review by philippmwirth — APPROVED** [review]
Looks good, let's merge this!
